src/Gui/DlgSettingsWorkbenchesImp.cpp

```diff
@@ -0,0 +1,475 @@
+    WorkbenchSwitcher::setIndex(index);
+}
+
+void DlgSettingsWorkbenchesImp::loadWorkbenchSelector()
+{
+    //workbench selector position combobox setup
+    ui->WorkbenchSelectorPosition->clear();
+    ui->WorkbenchSelectorPosition->addItem(tr("Toolbar"));
+    ui->WorkbenchSelectorPosition->addItem(tr("Left corner"));
+    ui->WorkbenchSelectorPosition->addItem(tr("Right corner"));
+    ui->WorkbenchSelectorPosition->setCurrentIndex(WorkbenchSwitcher::getIndex());
+}
+
+void DlgSettingsWorkbenchesImp::setStartWorkbenchComboItems()
+{
+    ui->AutoloadModuleCombo->clear();
+
+    // fills the combo box with activated workbenches.
+    QStringList enabledWbs;
+    for (int i = 0; i < ui->wbList->count(); i++) {
+        QWidget* widget = ui->wbList->itemWidget(ui->wbList->item(i));
+        if (widget) {
+            QCheckBox* enableCheckbox = widget->findChild<QCheckBox*>(enableCheckboxStr);
+            if (enableCheckbox && enableCheckbox->isChecked()) {
+                enabledWbs << widget->objectName();
+            }
+        }
+    }
+
+    QMap<QString, QString> menuText;
+    for (const auto& it : enabledWbs) {
+        QString text = Application::Instance->workbenchMenuText(it);
+        menuText[text] = it;
+    }
+
+    {   // add special workbench to selection
+        QPixmap px = Application::Instance->workbenchIcon(QString::fromLatin1("NoneWorkbench"));
+        QString key = QString::fromLatin1("<last>");
+        QString value = QString::fromLatin1("$LastModule");
+        if (px.isNull())
+            ui->AutoloadModuleCombo->addItem(key, QVariant(value));
+        else
+            ui->AutoloadModuleCombo->addItem(px, key, QVariant(value));
+    }
+
+    for (QMap<QString, QString>::Iterator it = menuText.begin(); it != menuText.end(); ++it) {
+        QPixmap px = Application::Instance->workbenchIcon(it.value());
+        if (px.isNull())
+            ui->AutoloadModuleCombo->addItem(it.key(), QVariant(it.value()));
+        else
+            ui->AutoloadModuleCombo->addItem(px, it.key(), QVariant(it.value()));
+    }
+
+    ui->AutoloadModuleCombo->setCurrentIndex(ui->AutoloadModuleCombo->findData(QString::fromStdString(_startupModule)));
+}
+
+void Gui::Dialog::DlgSettingsWorkbenchesImp::onStartWbChangedClicked(int index)
+{
+    //Update _startupModule
+    QVariant data = ui->AutoloadModuleCombo->itemData(index);
+    QString wbName = data.toString();
+    _startupModule = wbName.toStdString();
+
+    //Change wb that user can't deactivate.
+    for (int i = 0; i < ui->wbList->count(); i++) {
+        QWidget* widget = ui->wbList->itemWidget(ui->wbList->item(i));
+        if (widget) {
+            QCheckBox* enableCheckbox = widget->findChild<QCheckBox*>(enableCheckboxStr);
+            if (enableCheckbox && widget->objectName() == wbName) {
+                enableCheckbox->setEnabled(false);
+            }
+            else {
+                enableCheckbox->setEnabled(true);
+            }
+        }
+    }
```

Comment:
OK, got the answer to previous point. However such way of doing creates a strong interconnexion between widgets in the page which isn't ideal.

Reply:
Not sure what your point was here. Please have a look at the new code as it changed a lot with other fixes.